### PR TITLE
Feature/mongodb memory server set

### PR DESCRIPTION
**Pull request by Branchverse**
Implement memory server tests with 100% coverage and type tests in set module

2 Internal server errors excluded
1  development only endpoint excluded

### CODE REVIEW

File: src/app.module.ts [hunk 1]
```diff
@@ -1,36 +1,31 @@
...
+            imports: [ConfigModule],
+            useFactory: (configService: ConfigService) => ({
+                uri: configService.get<string>('DB_URI'),
+            }),
+            inject: [ConfigService],
+        }),
+        AuthModule,
+        UserModule,
+        SetModule,
+        //SystemModule,
```
Comment (coffemakingtoaster) [inline src/app.module.ts:25]: comment

File: src/set/set.controller.ts [hunk 2]
```diff
@@ -1,131 +1,118 @@
...
-  @ApiOperation({ summary: 'Get one Set by id' })
-  getOneSetFull(
-    @Param() { id }: MongoIdDto,
-    @Request() { user }: ParameterDecorator & { user: JwtUserDto }) {
-    throw new NotImplementedException('Currently disabled')
-    return //this.setService.getOneSetFull(id, user);
-  }
+    @Delete(':id')
+    @UseGuards(JwtAuthGuard)
+    // TODO MC: Can not reprocude error
```
Comment (coffemakingtoaster) [inline src/set/set.controller.ts:63]: is this still relevant? If not => remove

File: src/set/set.module.ts [hunk 3]
```diff
@@ -1,20 +1,18 @@
...
-    SharedModule,
-    MongooseModule.forFeature([{ name: 'Set', schema: SetSchema }]), 
-    MongooseModule.forFeature([{ name: 'Task', schema: TaskSchema }])
-  ],
-  controllers: [SetController],
-  providers: [SetService],
-  exports: [SetService]
+    // TODO MC: Does forFeature from Mongoose only import or instance? This could cause bugs..
+    imports: [
+        MongooseModule.forFeature([{ name: 'Set', schema: SetSchema }]),
```
Comment (coffemakingtoaster) [inline src/set/set.module.ts:11]: use Set.name and User.name

Comment (coffemakingtoaster) [reply src/set/set.module.ts:11]: or SetSchema.name? idk but that thing that is not a direct string

File: test/__mocks__/setMockData.ts [hunk 4]
```diff
@@ -1,32 +1,77 @@
...
+    return 'aaaaaaaaaaaaaaaaaaaaaaae'
+}
+export const getMockTask = () => {
+    return {
+        type: TaskType.TRUTH,
+        currentPlayerGender: CurrentPlayerGender.ANYONE,
+        message: 'new long enough Task',
+    }
+}
+export const getString = (number: number) => {
```
Comment (coffemakingtoaster) [inline test/__mocks__/setMockData.ts:71]: let str = new Array(len + 1).join( character ) 
use this this is less cluttered

File: src/set/set.service.ts [hunk 5]
```diff
@@ -1,345 +1,418 @@
...
-      }
-
-    } catch (error) {
-      if (error.code == '11000') {
-        throw new ConflictException('This set already exists')
-      }
-      console.error(error)
-      throw new InternalServerErrorException()
+    constructor(
+        @InjectModel('Set') private setSchema: Model<SetDocument>,
```
Comment (coffemakingtoaster) [inline src/set/set.service.ts:32]: Same thing with the strings...you might want to replace this

File: src/set/set.service.ts [hunk 6]
```diff
@@ -1,345 +1,418 @@
...
+            if (user.role != 'admin') throw new ForbiddenException()
 
-    return set
+            const set = await this.setSchema.findByIdAndDelete(id)
 
-  }
+            if (!set) throw new NotFoundException()
 
-  async deleteSet(id: ObjectId, deleteType: string, user: JwtUserDto): Promise<void> {
+            return
```
Comment (coffemakingtoaster) [inline src/set/set.service.ts:154]: while there is no rule written down anywhere the user endpoints return the deleted object....so maybe we want to find a way to unify this

File: src/set/set.service.ts [hunk 7]
```diff
@@ -1,345 +1,418 @@
...
-    if (!set)
-      throw new NotFoundException()
-
-    return
-
-  }
-
-  /*------------------------------------\
+    /*------------------------------------\
   |                Tasks                |
```
Comment (coffemakingtoaster) [inline src/set/set.service.ts:172]: these still look goofy as hell

File: src/set/set.service.ts [hunk 8]
```diff
@@ -1,345 +1,418 @@
...
+            ...set,
+            tasks: reducedTasks,
+        }
     }
-  }
 
-  // Uses 2 additional database calls to update the task counts and return the new settings
-  private async updateCounts(setId: ObjectId): Promise<UpdatedCounts> {
+    // Uses 2 additional database calls to update the task counts and return the new settings
+    private async updateCounts(setId: ObjectId): Promise<UpdatedCounts> {
```
Comment (coffemakingtoaster) [inline src/set/set.service.ts:310]: This probably is the best way to do this but this also is confusing. So either comment or create a short wiki entry to make understanding this easier :)

File: src/set/set.service.ts [hunk 9]
```diff
@@ -1,345 +1,418 @@
...
-          }
+                        type: task.type,
+                        currentPlayerGender: task.currentPlayerGender,
+                        message: task.message,
+                    },
+                    user,
+                )
+            })
+        })
+        // TODO: This is a security breach due to the mock data and the need to clean it in testing
```
Comment (coffemakingtoaster) [inline src/set/set.service.ts:409]: still valid after adding env guard?

File: test/set.e2e-spec.ts [hunk 10]
```diff
@@ -1,443 +1,765 @@
...
+                    previewImage: expect.any(String),
+                    bannerImage: expect.any(String),
+                }),
+            )
+        })
 
-  /*---------------------\ 
-  |         Sets         |
-  \---------------------*/
+        it('/set/:id (PATCH) by id by user', async () => {
```
Comment (coffemakingtoaster) [inline test/set.e2e-spec.ts:289]: maybe check in db as well?

File: test/set.e2e-spec.ts [hunk 11]
```diff
@@ -1,443 +1,765 @@
...
-          type: TaskType.DARE,
-          currentPlayerGender: exampleTask.currentPlayerGender,
-          message: exampleTask.message
-        })
-        .expect(HttpStatus.OK)
-      expect(res.body.daresCount === 1).toBeTruthy()
-      expect(res.body.truthCount === 0).toBeTruthy()
-    })
+        // Negative test
+        it('/set/:id (DELETE) by id with hard delete by admin', async () => {
```
Comment (coffemakingtoaster) [inline test/set.e2e-spec.ts:393]: this name does not sound like a negative test

**Review by coffemakingtoaster — COMMENTED** [review]
Some of this stuff is things we need to find a common solution for. 
For instance the return of a delete endpoint should be unified across all modules and endpoints

File: src/app.module.ts (hunk 1 above)
Comment (Branchverse) [reply src/app.module.ts:25]: Deleted

File: src/set/set.controller.ts (hunk 2 above)
Comment (Branchverse) [reply src/set/set.controller.ts:63]: This is an artifact from 200 years ago

File: src/set/set.module.ts (hunk 3 above)
Comment (Branchverse) [reply src/set/set.module.ts:11]: I will try thank you

File: src/set/set.service.ts (hunk 5 above)
Comment (Branchverse) [reply src/set/set.service.ts:32]: Yes

File: src/set/set.service.ts (hunk 6 above)
Comment (Branchverse) [reply src/set/set.service.ts:154]: Best practice is 204

File: src/set/set.service.ts (hunk 7 above)
Comment (Branchverse) [reply src/set/set.service.ts:172]: Is deleted

File: src/set/set.service.ts (hunk 9 above)
Comment (Branchverse) [reply src/set/set.service.ts:409]: Can be deleted on envGuard implementation, added TODO envGuard to find it later

Comment (Branchverse) [reply src/set/set.service.ts:409]: D

File: test/__mocks__/setMockData.ts (hunk 4 above)
Comment (Branchverse) [reply test/__mocks__/setMockData.ts:71]: Thank you, is done

File: test/set.e2e-spec.ts (hunk 11 above)
Comment (Branchverse) [reply test/set.e2e-spec.ts:393]: Fixed

File: test/set.e2e-spec.ts (hunk 10 above)
Comment (Branchverse) [reply test/set.e2e-spec.ts:289]: Done

**Comment by LarsFlieger** [comment]
lgtm, nice work!

**Review by LarsFlieger — APPROVED** [review]
lgtm, nice work! :)

**Comment by coffemakingtoaster** [comment]
lgtm

**Pull request merged**